source/cpp/korteweg-de-vries/KdV.cpp

```diff
@@ -0,0 +1,183 @@
+    adios2::ADIOS adios;
+    adios2::IO io = adios.DeclareIO("myio");
+    auto u_variable = io.DefineVariable<Real>(
+        "u", {size_t(N)}, {0}, {size_t(N)}, adios2::ConstantDims);
+    io.DefineAttribute<Real>("x0", 0);
+    io.DefineAttribute<Real>("dx", dx);
+    io.DefineAttribute<std::string>("interpretation", "Equispaced");
+    adios2::Engine adios_engine = io.Open("korteweg_de_vries.bp", adios2::Mode::Write);
+
+    std::vector<Real> u0(N);
+    for (int64_t i = 0; i < N; ++i)
+    {
+        u0[i] = cos(pi * i * dx);
+    }
+
+    adios_engine.BeginStep();
+    adios_engine.Put(u_variable, u0.data(), adios2::Mode::Sync);
+    adios_engine.EndStep();
+    Real original_momentum = momentum(u0);
+    std::vector<Real> u1(N);
+    for (int64_t i = 0; i < N; ++i)
+    {
+        Real cdt = cos(pi * i * dx) * dt;
+        u1[i] = cos(pi * (i * dx - cdt));
+    }
+
+    Real k1 = dt / (3 * dx);
+    Real k2 = delta * delta * dt / (dx * dx * dx);
+    Real t1, t2;
+    std::vector<Real> u2(N);
+    for (int64_t j = 1; j < M - 1; ++j)
+    {
+        t1 = (u1[1] + u1[0] + u1[N - 1]) * (u1[1] - u1[N - 1]);
+        t2 = u1[2] - 2 * u1[1] + 2 * u1[N - 1] - u1[N - 2];
+        u2[0] = u0[0] - k1 * t1 - k2 * t2;
+        t1 = (u1[2] + u1[1] + u1[0]) * (u1[2] - u1[0]);
+        t2 = u1[3] - 2 * u1[2] + 2 * u1[0] - u1[N - 1];
+        u2[1] = u0[1] - k1 * t1 - k2 * t2;
+        for (int64_t i = 2; i < N - 2; ++i)
+        {
+            t1 = (u1[i + 1] + u1[i] + u1[i - 1]) * (u1[i + 1] - u1[i - 1]);
+            t2 = u1[i + 2] - 2 * u1[i + 1] + 2 * u1[i - 1] - u1[i - 2];
+            u2[i] = u0[i] - k1 * t1 - k2 * t2;
+        }
+        u2[N - 2] =
+            u0[N - 2] -
+            k1 * (u1[N - 1] + u1[N - 2] + u1[N - 3]) * (u1[N - 1] - u1[N - 3]) -
+            k2 * (u1[0] - 2 * u1[N - 1] + 2 * u1[N - 3] - u1[N - 4]);
+        u2[N - 1] = u0[N - 1] -
+                    k1 * (u1[0] + u1[N - 1] + u1[N - 2]) * (u1[0] - u1[N - 2]) -
+                    k2 * (u1[1] - 2 * u1[0] + 2 * u1[N - 2] - u1[N - 3]);
+
+        Real p = momentum(u2);
+        if (std::abs(p) > std::sqrt(std::numeric_limits<Real>::epsilon()) ||
+            std::isnan(p))
+        {
+            std::cerr << "\nSolution diverged at t = " << (j + 1) * dt << "\n";
+            std::cerr << "Momentum = " << p << "\n";
+            return;
+        }
+        // The stability condition is very severe this iteration: We have to
+        // take way more time steps for stability than we need for accuracy.
+        // Hence we need skip a ton of steps so we don't spend all our time
+        // writing data:
+        int64_t skip_steps = 40000;
+        if ((j + 1) % skip_steps == 0)
+        {
+            display_progress(double(j + 1) / (M - 1));
+            adios_engine.BeginStep();
+            adios_engine.Put(u_variable, u2.data(), adios2::Mode::Sync);
+            adios_engine.EndStep();
+        }
+        u0 = u1;
+        u1 = u2;
+    }
```

Comment:
You need a call to Close. I'm surprised it works.

Reply:
@williamfgc : I thought I read in the docs that calls to close were optional; that's why I removed it.

Doesn't the destructor call Close() if not closed?

Reply:
Perhaps in non-MPI case. We can't use RAII after MPI Finalize. It is just good for portable MPI/no-MPI code and a good practice when dealing with C libraries (like MPI). It is up to you, but Close is more clear in the adios2 context since we are an MPI library.